### DISCUSSION TITLE
happy-dom support

### DISCUSSION THREAD

**Discussion by userquin** [antfu/vite-ssg#256] [General]
@blwtxc continue here

don't worry, it's complex and until you get used to the idea, it takes time

we should use a factory factory approach:
```ts
export interface SSGDOMBuildContext {
  document: Document
  serialize: () => Awaitable<string>
}
export interface SSGDOMBuildFactory {
  newSSGDOMContext: (renderedHTML: string): Awaitable<SSGBuildContext>
}
```

```ts
import { JSDOM } from 'jsdom'

export default jsdom()  {
  return {
    newSSGDOMContext(renderedHTML) {
      const jsdom= new JSDOM.JSDOM(renderedHTML)
      return {
        get document() { return jsdom.window.document },
        serialize() { return jsdom.serialize() },
      }
    }
  }
}
```

then on `src/node/build.ts`
```ts
  ...
  const queue = new PQueue.default({ concurrency })

  const { newSSGDOMContext }: SSGDOMBuildFactory = mock === 'happy-dom'
      ? await import('./happydom.mjs')
      : await import('./jsdom.mjs')

  for (const route of routesPaths) {
        ...
        // create jsdom from renderedHTML
        const jsdom = await newSSGDOMContext(renderedHTML)

        // render current page's preloadLinks
        renderPreloadLinks(jsdom.document, ctx.modules || new Set<string>(), ssrManifest)

        // render head
        head?.updateDOM(jsdom.document)

        const html = await jsdom.serialize()   
        ...
  }
  ...
```

**Comment by userquin** [comment]
https://github.com/capricorn86/happy-dom/blob/master/packages/happy-dom/src/xml-serializer/XMLSerializer.ts

**Comment by userquin** [comment]
@blwtxc you should also add the global for happy-dom, similar we do with jsdom, just checking for the mock: the main problem is you need to configure the globals, not just calling the register and unregister method

Anyway, I think the best approach is to use jsdom and to not include happy-dom, once merged my PR should work: it is only used for SSG and so jsdom should work

EDIT: I mean, you register only the happy-dom globals to make your app build without errors, if you're pacthing manuallly the build module just move the jsdom global before loading the ssr entry

**Comment by blwtxc** [comment]
Agreed. As I mentioned before: Sorry for wasting your time!

Cheers

Am 01.06.22 um 19:04 schrieb Joaquín Sánchez:
